src/Features/CSharp/Portable/UsePatternCombinators/AnalyzedPattern.cs

```diff
@@ -0,0 +1,136 @@
+        /// </summary>
+        internal abstract class Test : AnalyzedPattern
+        {
+            public readonly IOperation TargetOperation;
+
+            protected Test(IOperation targetOperation)
+                => TargetOperation = targetOperation;
+        }
+
+        /// <summary>
+        /// Represents a type-pattern, constructed from is-expression
+        /// </summary>
+        internal sealed class Type : Test
+        {
+            public readonly TypeSyntax TypeSyntax;
+
+            public Type(TypeSyntax expression, IOperation target) : base(target)
+                => TypeSyntax = expression;
+        }
+
+        /// <summary>
+        /// Represents a source-pattern, constructed from C# patterns
+        /// </summary>
+        internal sealed class Source : Test
+        {
+            public readonly PatternSyntax PatternSyntax;
+
+            public Source(PatternSyntax patternSyntax, IOperation target) : base(target)
+                => PatternSyntax = patternSyntax;
+        }
+
+        /// <summary>
+        /// Represents a constant-pattern, constructed from an equality check
+        /// </summary>
+        internal sealed class Constant : Test
+        {
+            public readonly ExpressionSyntax ExpressionSyntax;
+
+            public Constant(ExpressionSyntax expression, IOperation target) : base(target)
+                => ExpressionSyntax = expression;
+        }
+
+        /// <summary>
+        /// Represents a relational-pattern, constructed from relational operators
+        /// </summary>
+        internal sealed class Relational : Test
+        {
+            public readonly BinaryOperatorKind OperatorKind;
+            public readonly ExpressionSyntax Value;
+
+            public Relational(BinaryOperatorKind operatorKind, ExpressionSyntax value, IOperation target) : base(target)
+            {
+                OperatorKind = operatorKind;
+                Value = value;
+            }
+        }
+
+        /// <summary>
+        /// Represents an and/or pattern, constructed from a logical and/or expression.
+        /// </summary>
+        internal sealed class Binary : AnalyzedPattern
+        {
+            public readonly AnalyzedPattern Left;
+            public readonly AnalyzedPattern Right;
+            public readonly bool IsDisjunctive;
+            public readonly SyntaxToken Token;
+
+            private Binary(AnalyzedPattern leftPattern, AnalyzedPattern rightPattern, bool isDisjunctive, SyntaxToken token)
+            {
+                Left = leftPattern;
+                Right = rightPattern;
+                IsDisjunctive = isDisjunctive;
+                Token = token;
+            }
+
+            public static AnalyzedPattern Create(AnalyzedPattern leftPattern, AnalyzedPattern rightPattern, bool isDisjunctive, SyntaxToken token)
+            {
+                return !isDisjunctive && (leftPattern, rightPattern) is (Not left, Not right)
```

Comment:
why do we only unwrap in this one direction. i.e. `not a && not b   =>   not (a or b)`?  why not do the same for `not a || not b    =>     not (a and b)`?

Reply:
As currently speced, `not (a and b)` behavior would be "incorrect" from a logical prespective. We first match the operand and then negate the result, so `x is not (1 and 2)` is always true.

Reply:
On the second though, I guess we can do that. If we ended up with a nonsensical pattern, the original code had to be incorrect somehow.